src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import linear_algebra.matrix.orthogonal
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext i j,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h] }
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by { intros i j h, have h' := subsingleton.elim i j, contradiction }
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, rfl
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+λ i j, one_apply_ne
+
+@[simp] lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0) :
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+@[simp] lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.add
+  [add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A + B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.sub [add_group α]
+  {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
```

Comment:
```suggestion
lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
```

Reply:
Why did you delete the tag here? @jcommelin 
This lemma can close goals like `k \smul 1` automatically.

Reply:
How about keeping the tag for this one? @jcommelin

Reply:
@l534zhan I would prefer to remove it. `k \smul 1` should/could simp into `scalar k` and then a simp-lemma `is_diag_scalar` can close the goal. So this lemma doesn't need to be a simp-lemma in order to close such goals.